### PR TITLE
Require docstring for @public methods

### LINKED ISSUE #14996: Incorrect deprecation status in docs when a subclass overrides superclass deprecated property
### What's the issue or suggestion?

`AssetExecutionContext` and `OpExecutionContext` both specify an `asset_key_for_input` property. It is deprecated on `OpExecutionContext` (the superclass) but not on `AssetExecutionContext` the subclass. It should therefore show up as NOT deprecated for `AssetExecutionContext`, but it is showing up as deprecated.

(This will be fixed by requiring docstrings for `@public`, i.e. docs-rendered, methods-- see #15078).

### Additional information

_No response_

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

### LINKED ISSUE #15076: All `@public` methods should require docstrings.
### What's the issue or suggestion?

Only `@public` methods appear in Dagster's docs. We should require docstrings on all such methods and fail the docs build when they aren't present.

Not only does this provide a better docs experience, but it also prevents weird issues like #14996 which arise because of how sphinx differently processes objects having docstrings vs those missing them.

### Additional information

_No response_

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

- Resolves #14996.
- Resolves #15076.

This updates our custom sphinx extension so that docstrings are required for `@public` methods. The docs build will fail if a `@public` method with no docstring is encountered.

This will enhance the quality of our docs and prevent strange behaviors that can occur in sphinx due to differential handling of objects with and without docstrings.

(NOTE: I also tacked the last few straggler docstring additions into this PR after the below list was completed)

---

Before this can land, there are about 200 `@public` methods that need docstrings. If we split the work it shouldn't be too bad. I have split the missing docstrings into ~10-ish groups below. If you want to help with this project, please comment and volunteer to take on 1-2 groups-- I'll check off the ones that are taken. I'll start by taking `DagsterInstance` and `Misc`.

@sryza @clairelin135 @OwenKephart @dpeng817 @benpankow @jamiedemaria @yuhan @tacastillo @alangenfeld 

(just names that came to mind that might be interested, no obligation)

- [x] Assets & Ops (@OwenKephart)
    - AssetsDefinition.can_subset
    - AssetsDefinition.dependency_keys
    - AssetsDefinition.descriptions_by_key
    - AssetsDefinition.get_partition_mapping
    - AssetsDefinition.group_names_by_key
    - AssetsDefinition.key
    - AssetsDefinition.keys
    - AssetsDefinition.node_def
    - AssetsDefinition.op
    - AssetsDefinition.partitions_def
    - AssetsDefinition.required_resource_keys
    - AssetsDefinition.resource_defs
    - OpDefinition.alias
    - OpDefinition.config_schema
    - OpDefinition.ins
    - OpDefinition.name
    - OpDefinition.outs
    - OpDefinition.required_resource_keys
    - OpDefinition.retry_policy
    - OpDefinition.tag
    - OpDefinition.tags
    - OpDefinition.version
    - OpDefinition.with_hooks
    - OpDefinition.with_retry_policy

- [x] Config & Resources (@benpankow)
    - Field.default_value
    - Field.description
    - Field.is_required
    - Map.key_label_name
    - Array.description
    - ResourceDefinition.description
    - ResourceDefinition.required_resource_keys
    - ResourceDefinition.string_resource
    - ResourceDefinition.version
    - InitResourceContext.instance
    - InitResourceContext.log
    - InitResourceContext.resources

- [x] DagsterInstance (@smackesey)
    - DagsterInstance.add_dynamic_partitions
    - DagsterInstance.delete_dynamic_partition
    - DagsterInstance.delete_run
    - DagsterInstance.ephemeral
    - DagsterInstance.get
    - DagsterInstance.get_asset_keys
    - DagsterInstance.get_asset_records
    - DagsterInstance.get_dynamic_partitions
    - DagsterInstance.get_event_records
    - DagsterInstance.get_latest_materialization_event
    - DagsterInstance.get_run_by_id
    - DagsterInstance.get_run_record_by_id
    - DagsterInstance.get_run_records
    - DagsterInstance.has_asset_key
    - DagsterInstance.has_dynamic_partition
    - DagsterInstance.local_temp
    - DagsterInstance.wipe_assets

- [x] Dagster Types (@smackesey)
    - DagsterType.description
    - DagsterType.has_unique_name
    - DagsterType.loader
    - DagsterType.required_resource_keys
    - DagsterType.type_check
    - DagsterType.typing_type

- [x] Inputs & Outputs (@jamiedemaria) 
    - DynamicOutput.mapping_key
    - DynamicOutput.output_name
    - DynamicOutput.value
    - OutputContext.asset_key
    - OutputContext.get_asset_identifier
    - OutputContext.has_asset_key
    - OutputContext.has_asset_partitions
    - InputContext.asset_key
    - InputContext.get_asset_identifier
    - InputContext.has_asset_key
    - InputContext.has_asset_partitions

- [x] Graphs & Jobs (@dpeng817)
    - GraphDefinition.alias
    - GraphDefinition.config_mapping
    - GraphDefinition.input_mappings
    - GraphDefinition.name
    - GraphDefinition.output_mappings
    - GraphDefinition.tag
    - GraphDefinition.tags
    - GraphDefinition.with_hooks
    - GraphDefinition.with_retry_policy
    - JobDefinition.config_mapping
    - JobDefinition.executor_def
    - JobDefinition.has_specified_executor
    - JobDefinition.has_specified_loggers
    - JobDefinition.loggers
    - JobDefinition.partitioned_config
    - JobDefinition.partitions_def
    - JobDefinition.resource_defs

- [x] Misc (@smackesey)
    - OpExecutionContext.op_config
    - MultiDependencyDefinition.get_dependencies_and_mappings
    - MultiDependencyDefinition.get_node_dependencies
    - MultiDependencyDefinition.is_fan_in
    - HookContext.instance
    - ExecutorDefinition.executor_creation_fn
    - DagstermillExecutionContext.job_name
    - LoggerDefinition.config_schema
    - LoggerDefinition.description
    - LoggerDefinition.logger_fn
    - VersionStrategy.get_op_version
    - VersionStrategy.get_resource_version
    - SourceHashVersionStrategy.get_op_version
    - SourceHashVersionStrategy.get_resource_version
    - ExecuteInProcessResult.asset_value
    - ExecuteInProcessResult.dagster_run
    - ExecuteInProcessResult.job_def
    - ExecuteInProcessResult.run_id

- [x] Metadata (@smackesey)
    - MetadataValue.value
    - DagsterAssetMetadataValue.value
    - JsonMetadataValue.value
    - PathMetadataValue.value
    - NotebookMetadataValue.value
    - PythonArtifactMetadataValue.value
    - TableMetadataValue.infer_column_type
    - TableMetadataValue.value
    - TextMetadataValue.value
    - UrlMetadataValue.value

- [x] Partitions (@clairelin135)
    - PartitionedConfig.get_partition_keys
    - PartitionedConfig.partitions_def
    - PartitionedConfig.run_config_for_partition_fn
    - PartitionedConfig.run_config_for_partition_key_fn
    - PartitionedConfig.tags_for_partition_fn
    - PartitionedConfig.tags_for_partition_key_fn
    - PartitionsDefinition.get_partition_keys
    - TimeWindowPartitionsDefinition.day_offset
    - TimeWindowPartitionsDefinition.hour_offset
    - TimeWindowPartitionsDefinition.minute_offset
    - TimeWindowPartitionsDefinition.schedule_type
    - StaticPartitionsDefinition.get_partition_keys
    - MultiPartitionsDefinition.get_partition_keys
    - DynamicPartitionsDefinition.get_partition_keys

- [x] Repositories (@smackesey)
    - RepositoryDefinition.description
    - RepositoryDefinition.get_schedule_def
    - RepositoryDefinition.get_sensor_def
    - RepositoryDefinition.has_schedule_def
    - RepositoryDefinition.has_sensor_def
    - RepositoryDefinition.name
    - RepositoryDefinition.schedule_defs
    - RepositoryDefinition.sensor_defs
    - RepositoryData.get_all_sensors
    - RepositoryData.get_assets_defs_by_key
    - RepositoryData.get_sensor
    - RepositoryData.get_sensor_names
    - RepositoryData.get_source_assets_by_key
    - RepositoryData.has_sensor

- [x] Runs & Events (@OwenKephart)
    - DagsterEvent.asset_key
    - DagsterEvent.is_alert_event
    - DagsterEvent.is_asset_materialization_planned
    - DagsterEvent.is_asset_observation
    - DagsterEvent.is_engine_event
    - DagsterEvent.is_expectation_result
    - DagsterEvent.is_failure
    - DagsterEvent.is_handled_output
    - DagsterEvent.is_hook_event
    - DagsterEvent.is_loaded_input
    - DagsterEvent.is_resource_init_failure
    - DagsterEvent.is_step_event
    - DagsterEvent.is_step_failure
    - DagsterEvent.is_step_materialization
    - DagsterEvent.is_step_restarted
    - DagsterEvent.is_step_skipped
    - DagsterEvent.is_step_start
    - DagsterEvent.is_step_success
    - DagsterEvent.is_step_up_for_retry
    - DagsterEvent.is_successful_output
    - DagsterEvent.partition
    - DagsterRun.is_failure
    - DagsterRun.is_failure_or_canceled
    - DagsterRun.is_finished
    - DagsterRun.is_resume_retry
    - DagsterRun.is_success
    - EventLogEntry.dagster_event_type
    - EventLogEntry.get_dagster_event
    - EventLogEntry.is_dagster_event

- [x] Schedules & Sensors (@OwenKephart)
    - ScheduleDefinition.cron_schedule
    - ScheduleDefinition.default_status
    - ScheduleDefinition.description
    - ScheduleDefinition.environment_vars
    - ScheduleDefinition.execution_timezone
    - ScheduleDefinition.job
    - ScheduleDefinition.job_name
    - ScheduleDefinition.name
    - ScheduleDefinition.required_resource_keys
    - ScheduleEvaluationContext.instance
    - SensorDefinition.default_status
    - SensorDefinition.description
    - SensorDefinition.job
    - SensorDefinition.job_name
    - SensorDefinition.jobs
    - SensorDefinition.minimum_interval_seconds
    - SensorDefinition.name
    - SensorDefinition.required_resource_keys
    - SensorEvaluationContext.instance
    - SensorEvaluationContext.last_completion_time
    - SensorEvaluationContext.last_run_key
    - SensorEvaluationContext.repository_def
    - SensorEvaluationContext.repository_name
    - SensorEvaluationContext.resources
    - AssetSensorDefinition.asset_key
    - MultiAssetSensorEvaluationContext.asset_keys
    - MultiAssetSensorEvaluationContext.assets_defs_by_key
    - RunStatusSensorContext.partition_key

## How I Tested These Changes

Built docs.

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #15077** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15077" target="_blank">Graphite</a> 
    * **PR #15078** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15078" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/15078?utm_source=stack-comment).

**Comment by dpeng817** [comment]
I'll take graphs and jobs

**Comment by jamiedemaria** [comment]
i can take Inputs and Outputs to start and then maybe another once i get that done

**Comment by rexledesma** [comment]
Love this change -- is it possible to enforce that these public methods also have attached examples for usage?

**Comment by smackesey** [comment]
>Love this change -- is it possible to enforce that these public methods also have attached examples for usage?

This is a little tricky because examples are a lot of work and there are many public methods (especially properties/getters) where the examples would be trivial and add little to no value.

But I do think there's definite value in requiring examples for methods over some level of complexity-- the question is how to define it. One possibility is to require examples for methods with at least N args-- intuitively I would probably set N at 3.

Aside from examples, the next step I had in mind for this project was enforcing that non-private parameters in public methods are documented.

**Comment by rexledesma** [comment]
For integrations, where the surface area is smaller, we should be enforcing examples. If we could selectively enforce it on high priority integrations, I would be happy for now.

**Comment by sryza** [comment]
Want to assign me a sub-area? Maybe one of the many ones that are currently assigned to @OwenKephart ?

**Comment by smackesey** [comment]
>Want to assign me a sub-area?

@sryza They're all done! This is good to merge.

File: docs/sphinx/conf.py [hunk 1]
```diff
@@ -149,6 +149,7 @@
     "mlflow",
     "mysql",
     "oauth2client",
+    "orjson",
```
Comment (sryza) [inline docs/sphinx/conf.py:152]: is this a related change?

**Review by sryza — APPROVED** [review]
awesome

File: docs/sphinx/conf.py (hunk 1 above)
Comment (smackesey) [reply docs/sphinx/conf.py:152]: Related enough-- has to do with docs build, which started failing on this PR due to this new dep.
